### PR TITLE
[Fix-15992] maintain slot index by iterate MasterPriorityQueue orderly

### LINKED ISSUE #15992: [Bug] [dolphinscheduler-service] iterator order of MasterPriorityQueue is undefined
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

According to PriorityBlockingQueue document, iterator order is undefined.

> This class and its iterator implement all of the optional methods of the Collection and Iterator interfaces. The Iterator provided in method iterator() and the Spliterator provided in method `iterator()` are not guaranteed to traverse the elements of the PriorityBlockingQueue in any particular order. If you need ordered traversal, consider using `Arrays.sort(pq.toArray())`. 

In dolphinscheduler operation, Every `WorkerGroupRefreshInterval`,   `ServerNodeManager.refreshNodesAndGroupMappings` will  be triggerred, and the master slot index will be refreshed by  `refreshMasterList`.

But iterator order of MasterPriorityQueue is undefined, thus even if the host server create time is not affected, the slot index may change every `WorkerGroupRefreshInterval`.  This will make slot assigment between servers change unintentionally.

```java
private void refreshMasterList() {
        hostIndexMap.clear();
        Iterator<Server> iterator = queue.iterator();
        int index = 0;
        while (iterator.hasNext()) {
            Server server = iterator.next();
            String addr = NetUtils.getAddr(server.getHost(), server.getPort());
            hostIndexMap.put(addr, index);
            index += 1;
        }
    }
```

```java
    class WorkerNodeInfoAndGroupDbSyncTask implements Runnable {

        @Override
        public void run() {
            try {
                // sync worker node info
                refreshNodesAndGroupMappings();
            } catch (Exception e) {
                log.error("WorkerNodeInfoAndGroupDbSyncTask error:", e);
            }
        }
    }

```

```java
    /**
     * Refresh master/worker nodes and worker group mapping information
     */
    private void refreshNodesAndGroupMappings() {
        updateWorkerNodes();
        updateWorkerGroupMappings();
        notifyWorkerInfoChangeListeners();

        updateMasterNodes();
        notifyMasterInfoChangeListeners();
    }
```

### What you expected to happen

If not server is added or removed, the slot index should be the same guranteed.

### How to reproduce

According to document, the iteration order  is not guranteed.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by coldestlin**
## Purpose of the pull request

fix https://github.com/apache/dolphinscheduler/issues/15992

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java [hunk 1]
```diff
@@ -80,15 +81,14 @@ public void clear() {
...
         int index = 0;
-        while (iterator.hasNext()) {
-            Server server = iterator.next();
+        for (Server server: nQueue) {
             String addr = NetUtils.getAddr(server.getHost(), server.getPort());
             hostIndexMap.put(addr, index);
             index += 1;
         }
-
     }
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java]: It's needed to add a private method to get the Elements from queue, and add UT for this.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java (hunk 1 above)
Comment (coldestlin) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java]: modified, and UT added, please have a look

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java [hunk 2]
```diff
@@ -80,15 +80,23 @@ public void clear() {
...
             index += 1;
         }
+    }
 
+    /**
+     * get ordered collection of priority queue
+     *
+     * @return ordered collection
+     */
+    private Server[] getOrderedElements() {
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java]:
```suggestion
    Server[] getServers() {
```

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueueTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        // Test empty queue
+        Server[] emptyElements = (Server[]) getOrderedElementsMethod.invoke(queue);
+        Assertions.assertArrayEquals(emptyElements, new Server[]{});
+
+        // Test queue with fabricated servers
+        queue.putAll(getServerList());
+        Server[] orderElements = (Server[]) getOrderedElementsMethod.invoke(queue);
+        Assertions.assertEquals(extractServerIds(orderElements), Arrays.asList(4, 2, 1, 3));
+
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueueTest.java]: Don't use reflection to test private methods

Comment (coldestlin) [reply dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueueTest.java]: Got it, modified.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java (hunk 2 above)
Comment (coldestlin) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/queue/MasterPriorityQueue.java]: Accepted

**Comment by ruanwenjun** [comment]
Please fix the code style by `mvn spotless:apply`

**Review by rickchengx — APPROVED** [review]
LGTM

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Pull request merged**